lib/mlpatches/modules/subprocess.py

```diff
@@ -0,0 +1,301 @@
+# -*- coding: utf-8 -*-
+"""Subprocess management"""
+# info: parts of this module has been copied from the original subprocess-module
+
+# === NotImplemented ===
+# - the following args are always ignored:
+# - executable
+# - shell
+# - startupinfo
+# - preexec_fn
+# - bufsize
+# - creationflags
+# - close_fds
+# Other implemention differences:
+# - terminate() is just a kill()
+# - send_signal() always call kill()
+# - communicate() splits data in chunks of 4096 bytes before sending
+# - poll() never returns the signal that killed the process
+# - pid is the job_id
+
+import os
+import select
+import time
+import sys
+from mlpatches import base, l2c
+from six import integer_types, string_types
+
+_stash = base._stash
+list2cmdline = l2c.list2cmdline
+
+# constants
+try:
+    # setup MAXFD. we don't require this, but other scripts may expect this value to be present
+    MAXFD = os.sysconf("SC_OPEN_MAX")
+except:
+    MAXFD = 256
+
+PIPE = -1
+STDOUT = -2
+
+
+class CalledProcessError(Exception):
+    """Exception raised when a process run by check_call() or check_output() returns a non-zero exit status."""
+
+    def __init__(self, returncode, cmd, output=None):
+        self.returncode = returncode
+        self.cmd = cmd
+        self.output = output
+
+    def __str__(self):
+        return "Command '{c}' returned non-zero exit status {s}".format(
+            c=self.cmd, s=self.returncode
+        )
+
+
+def call(*args, **kwargs):
+    """Run the command described by args. Wait for command to complete, then return the returncode attribute."""
+    return Popen(*args, **kwargs).wait()
+
+
+def check_call(*args, **kwargs):
+    """Run command with arguments. Wait for command to complete. If the return code was zero then return, otherwise raise CalledProcessError. The CalledProcessError object will have the return code in the returncode attribute."""
+    rc = call(*args, **kwargs)
+    if rc != 0:
+        cmd = kwargs.get("args")
+        if cmd is None:
+            cmd = args[0]
+        raise CalledProcessError(rc, cmd)
+    return 0
+
+
+def check_output(*args, **kwargs):
+    """Run command with arguments and return its output as a byte string.
+
+    If the return code was non-zero it raises a CalledProcessError. The CalledProcessError object will have the return code in the returncode attribute and any output in the output attribute."""
+    if "stdout" in kwargs:
+        raise ValueError("stdout argument not allowed, it will be override.")
+    p = Popen(stdout=PIPE, *args, **kwargs)
+    out, _ = p.communicate()
+    rc = p.poll()
+    if rc != 0:
+        cmd = kwargs.get("args")
+        if cmd is None:
+            cmd = args[0]
```

Comment:
Optional simplification.
```suggestion
        cmd = kwargs.get("args") or args[0]
```

Reply:
should I add it to this PR?
This code was not modified by me

Reply:
Nope.